quic/transport/ngtcp2/stream/openstate.nim

```diff
@@ -31,28 +33,23 @@ method enter*(state: OpenStream, stream: Stream) =
 method leave*(state: OpenStream) =
   procCall leave(StreamState(state))
   state.stream = Opt.none(Stream)
+  # TODO: clear userdata
 
 method read*(state: OpenStream): Future[seq[byte]] {.async.} =
   let incomingFut = state.incoming.get()
-  let timeoutFut = state.connection.timeout.expired()
-  let raceFut = await race(incomingFut, state.cancelRead, timeoutFut)
-  if (await race(incomingFut, state.cancelRead)) == incomingFut:
+  let raceFut = await race(state.closeFut, incomingFut)
+  if raceFut == incomingFut:
     result = await incomingFut
     allowMoreIncomingBytes(state.stream, state.connection, result.len.uint64)
   else:
     incomingFut.cancelSoon()
     let stream = state.stream.valueOr:
       return
     if state.frameSorter.isEOF():
-      stream.switch(
-        newClosedStream(state.incoming, state.frameSorter, state.connection)
-      )
+      stream.switch(newClosedStream(state.incoming, state.frameSorter))
 
-    raise
-      if raceFut == timeoutFut:
-        newException(StreamError, "stream timed out")
-      else:
-        newException(StreamError, "stream is closed")
+    let closeReason = await state.closeFut
+    raise newException(StreamError, closeReason)
```

Comment:
this error might as well be `ClosedStreamError`. that is, all closed errors, from any stream state should ideally be `ClosedStreamError`

Reply:
good catch!